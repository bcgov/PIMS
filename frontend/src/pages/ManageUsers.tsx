import * as React from 'react';
import { Container, Row, Col } from 'react-bootstrap';

const ManageUsers = () => {
  return (
    <Container fluid={true}>
      <Row>
        <Col>
          <h1>Manage Users</h1>
        </Col>
      </Row>
    </Container>
  );
};

export default ManageUsers;
