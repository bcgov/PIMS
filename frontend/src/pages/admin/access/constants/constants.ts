import { ColDef } from 'ag-grid-community';
import { IAccessRequest } from 'interfaces';
import { AccessStatusDisplay, AccessStatusDisplayMapper } from 'constants/accessStatus';

export const columnDefinitions = (): ColDef[] => {
  let items: ColDef[] = [
    {
      headerName: 'IDIR/BCeID',
      checkboxSelection: true,
      headerCheckboxSelection: true,
      valueGetter: params => (params.data as IAccessRequest).user.username,
    },
    {
      headerName: 'First name',
      valueGetter: params => (params.data as IAccessRequest).user.firstName,
    },
    {
      headerName: 'Last name',
      valueGetter: params => (params.data as IAccessRequest).user.lastName,
    },
    {
      headerName: 'Email',
      valueGetter: params => (params.data as IAccessRequest).user.email,
      width: 250,
    },
    {
      headerName: 'Position',
      valueGetter: params => (params.data as IAccessRequest).position,
    },
    {
      headerName: 'Status',
      valueGetter: (params): AccessStatusDisplay =>
        AccessStatusDisplayMapper[(params.data as IAccessRequest).status],
      filter: false,
      sortable: false,
    },
    {
      headerName: 'Agency',
      valueGetter: params => (params.data as IAccessRequest).agencies[0].name,
      width: 250,
    },
    {
      headerName: 'Role',
      valueGetter: params => (params.data as IAccessRequest).roles[0].name,
    },
  ];

  items = items.map(column => ({
    filter: true,
    suppressMenu: true,
    cellClass: 'user-info-cell',
    sortable: true,
    minWidth: 230,
    ...column,
  }));

  return items;
};
