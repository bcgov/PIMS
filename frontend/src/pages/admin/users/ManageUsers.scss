.user-management-table-wrapper {
  > th {
    font-weight: bold;
  }
}

div[class^='MUIDataTableHeadCell-data-'] {
  font-weight: bold;
}
.MuiTableHead-root {
  tr {
    th {
      background-color: lighten($color: grey, $amount: 40%) !important;
    }
  }
}

div[class^='MUIDataTableToolbar-titleRoot'] {
  text-align: left;
  h6 {
    font-weight: bolder !important;
    font-size: 20px;
  }
}
.MuiTableRow-root.Mui-selected,
.MuiTableRow-root.Mui-selected:hover {
  background-color: lighten($color: #38598a, $amount: 50%) !important;
}
