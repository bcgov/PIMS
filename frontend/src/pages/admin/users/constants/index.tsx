import * as React from 'react';
import { MUIDataTableColumnDef } from 'mui-datatables';
import { FaCheck, FaTimes } from 'react-icons/fa';
import { AccountActive } from '../interfaces/IUserRecord';
import { RowActions } from '../components/RowActions';
export const columns: MUIDataTableColumnDef[] = [
  {
    name: 'id',
    label: 'ID',
    options: {
      display: 'excluded',
      filter: false,
    },
  },
  {
    name: 'username',
    label: 'IDIR/BCeID',
    options: {
      customBodyRender: (username: string, tableMeta: any) => {
        return <a href={`/admin/user/${tableMeta.rowData[0]}`}>{username}</a>;
      },
    },
  },
  {
    name: 'firstName',
    label: 'First name',
  },
  {
    name: 'lastName',
    label: 'Last name',
  },
  {
    name: 'email',
    label: 'Email',
  },
  {
    name: 'active',
    label: 'Active',
    options: {
      customBodyRender: (active: AccountActive, tableMeta: any) => {
        return active === AccountActive.YES ? <FaCheck /> : <FaTimes />;
      },
    },
  },
  {
    name: 'agency',
    label: 'Agency',
  },
  {
    name: 'position',
    label: 'Position',
  },
  {
    name: 'role',
    label: 'Role',
  },
  {
    name: 'action',
    label: 'Actions',
    options: {
      customBodyRender: (value: any, tableMeta: any) => {
        return <RowActions userId={tableMeta.rowData[0]} active={tableMeta.rowData[5]} />;
      },
    },
  },
];
