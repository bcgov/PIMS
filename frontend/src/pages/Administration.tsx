import * as React from 'react';
import { Container, Row, Col, Nav } from 'react-bootstrap';
import { useHistory, useRouteMatch, Switch, Route } from 'react-router-dom';
import ManageUsers from './ManageUsers';
import ManageAccessRequests from './ManageAccessRequests';

const Administration = () => {
  const history = useHistory();
  let { path, url } = useRouteMatch();
  const handleSelect = (eventKey: string) => history.push(`${url}/${eventKey}`);

  React.useEffect(() => {
    handleSelect('users');
  }, []);

  return (
    <Container fluid={true}>
      <Row>
        <Col>
          <h1>Administrator Dashboard</h1>
        </Col>
      </Row>
      <Row>
        <Col>
          <Nav variant="tabs" defaultActiveKey="/home" onSelect={handleSelect}>
            <Nav.Item>
              <Nav.Link eventKey="users">Manage Users</Nav.Link>
            </Nav.Item>
            <Nav.Item>
              <Nav.Link eventKey="requests">Manage Access Requests</Nav.Link>
            </Nav.Item>
          </Nav>
        </Col>
      </Row>
      <Row>
        <Col>
          <Switch>
            <Route path={`${path}/users`} component={ManageUsers}></Route>
            <Route path={`${path}/requests`} component={ManageAccessRequests}></Route>
          </Switch>
        </Col>
      </Row>
    </Container>
  );
};

export default Administration;
