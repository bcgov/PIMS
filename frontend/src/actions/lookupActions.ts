import * as ActionTypes from 'constants/actionTypes';

//Parcel List API action

export interface ILookupCode {
  code: string;
  name: string;
  id: string;
  isDisabled: boolean;
  isPublic?: boolean;
  type: string;
  parentId?: number;
}

export interface IStoreLookupCodeAction {
  type: typeof ActionTypes.STORE_LOOKUP_CODE_RESULTS;
  lookupCodes: ILookupCode[];
}

export const storeLookupCodesAction = (lookupCodes: ILookupCode[]) => ({
  type: ActionTypes.STORE_LOOKUP_CODE_RESULTS,
  lookupCodes: lookupCodes,
});
