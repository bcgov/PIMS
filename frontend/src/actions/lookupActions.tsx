import * as ActionTypes from 'constants/actionTypes';

//Parcel List API action

export interface ILookupCode {
  code: string;
  name: string;
  id: string;
  isDisabled: boolean;
  type: string;
}

export interface IStoreLookupCodeAction {
  type: typeof ActionTypes.STORE_LOOKUP_CODE_RESULTS;
  lookupCodes: ILookupCode[];
}

export const storeLookupCodesAction = (lookupCodes: ILookupCode[]) => ({
  type: ActionTypes.STORE_LOOKUP_CODE_RESULTS,
  lookupCodes: lookupCodes,
});
