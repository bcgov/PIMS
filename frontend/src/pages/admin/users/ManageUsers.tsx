import React, { useEffect } from 'react';
import { Container, Spinner } from 'react-bootstrap';
import { getUsersAction } from 'actionCreators/usersActionCreator';
import { useDispatch, useSelector } from 'react-redux';
import { RootState } from 'reducers/rootReducer';
import { IStoreUsersAction } from 'actions/adminActions';
import { IPagedItems } from 'interfaces/pagedItems';
import { toApiPaginateParams } from 'utils/CommonFunctions';
import { IGenericNetworkAction } from 'actions/genericActions';
import * as actionTypes from 'constants/actionTypes';
import './ManageUsers.scss';
import { UsersList, TableStateChangeFn } from './components/UsersList';
import { IUser } from 'interfaces';
import { MUIDataTableState } from 'mui-datatables';
import { IUserRecord, AccountActive } from './interfaces/IUserRecord';

const ManageUsers = () => {
  const dispatch = useDispatch();
  const MAX_USERS_PER_PAGE = 10;
  const pagedUsers = useSelector<RootState, IPagedItems<IUser>>(state => {
    return (state.users as IStoreUsersAction).pagedUsers;
  });
  const users = useSelector<RootState, IGenericNetworkAction>(
    state => (state.network as any)[actionTypes.GET_USERS] as IGenericNetworkAction,
  );
  useEffect(() => {
    dispatch(getUsersAction(toApiPaginateParams(0, MAX_USERS_PER_PAGE)));

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const tableChange: TableStateChangeFn = (type: string, state: MUIDataTableState) => {
    switch (type) {
      case 'pagination':
        break;
      case 'filter':
        break;
      case 'sort':
        break;
    }
  };
  const userList = pagedUsers.items.map(
    (u: IUser): IUserRecord => ({
      id: u.id,
      email: u.email,
      username: u.username,
      firstName: u.firstName,
      lastName: u.lastName,
      active: u.isDisabled ? AccountActive.NO : AccountActive.YES,
      role: u.roles ? u.roles[0].name : 'No role attached',
      agency: u.agencies ? u.agencies[0].name : 'No agency associated',
      position: u.position,
    }),
  );

  return users && !users.isFetching ? (
    <Container fluid={true}>
      <UsersList
        data={userList}
        page={pagedUsers.page}
        pageSize={MAX_USERS_PER_PAGE}
        totalRows={pagedUsers.total}
        onTableChange={tableChange}
      />
    </Container>
  ) : (
    <Spinner animation="border"></Spinner>
  );
};

export default ManageUsers;
