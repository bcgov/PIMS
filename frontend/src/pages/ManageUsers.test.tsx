import React from 'react';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme, { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Administration from './Administration';
import { ILookupCode } from 'actions/lookupActions';
import * as actionTypes from 'constants/actionTypes';
import * as reducerTypes from 'constants/reducerTypes';
import * as API from 'constants/API';
import { render, fireEvent, queryByText } from '@testing-library/react';
import ManageUsers from './ManageUsers';
import { FormikLookupCodeDropdown } from 'components/common/LookupCodeDropdown';

Enzyme.configure({ adapter: new Adapter() });

const history = createMemoryHistory();
history.push('admin');
const mockStore = configureMockStore([thunk]);

const lCodes = {
  lookupCodes: [
    { name: 'agencyVal', id: '1', isDisabled: false, type: API.AGENCY_CODE_SET_NAME },
    { name: 'roleVal', id: '1', isDisabled: false, type: API.ROLE_CODE_SET_NAME },
  ] as ILookupCode[],
};

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'), // use actual for all non-hook parts
  useRouteMatch: () => ({ url: '/admin', path: '/admin' }),
}));

// Empty response
const store = mockStore({
  [reducerTypes.USERS]: { pagedUsers: [] },
  [reducerTypes.LOOKUP_CODE]: lCodes,
  [reducerTypes.NETWORK]: { [actionTypes.GET_USERS]: {} },
});
const successStore = mockStore({
  [reducerTypes.USERS]: {
    pagedUsers: {
      page: 1,
      total: 2,
      quantity: 2,
      items: [
        {
          id: 1,
          displayName: 'testUser',
          agencies: [{ id: '1', name: 'HLTH' }],
          roles: [{ id: '1', name: 'admin' }],
        },
      ],
    },
  },
  [reducerTypes.LOOKUP_CODE]: lCodes,
  [reducerTypes.NETWORK]: { [actionTypes.GET_USERS]: {} },
});
const loadingStore = mockStore({
  [reducerTypes.USERS]: {},
  [reducerTypes.LOOKUP_CODE]: lCodes,
  [reducerTypes.NETWORK]: { [actionTypes.GET_USERS]: { isFetching: true } },
});

it('renders Manage Users page correctly', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Router history={history}>
          <ManageUsers />
        </Router>
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

describe('component functionality', () => {
  const componentRender = (store: any) =>
    render(
      <Provider store={store}>
        <Router history={history}>
          <ManageUsers />
        </Router>
      </Provider>,
    );

  it('renders no users by default', () => {
    const { getByText } = componentRender(store);
    expect(getByText(/No Users available to be managed/i));
  });

  it('renders nothing while loading', () => {
    const { queryByText } = componentRender(loadingStore);
    expect(queryByText(/Manage Users/i)).toBeNull();
  });

  it('renders a table with users and their associated values', () => {
    const { getByText } = componentRender(successStore);
    expect(getByText(/testUser/i));
    expect(getByText(/HLTH/i));
    expect(getByText(/admin/i));
  });

  it('allows useres the option to edit', () => {
    const { getByText } = componentRender(successStore);
    expect(getByText(/edit/i));
    //
  });
});
