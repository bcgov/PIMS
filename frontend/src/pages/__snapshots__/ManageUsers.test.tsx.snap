// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Manage Users page correctly 1`] = `
<div
  className="container-fluid"
>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <h2>
        Manage Users
      </h2>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <p>
        No Users available to be managed
      </p>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col"
    />
  </div>
</div>
`;
